<a href="https://colab.research.google.com/github/chadschlett/sports_analytics/blob/master/cschlett_assignment2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Chadwick Schlett University of Michigan Graduate Student

Visualization Technique:

The visualization I plan to use is a box-plot showing passing yards data for Tom Brady over the course of his NFL career. The Box-plot will show a year-by-year analysis of Brady's games highlighting the median yards per game for that season as well as the first and third quartiles. Furthermore, the Max and Minimum values will also be shown per year. I will also add a subplot that shows average yards a game. These plots give the viewer a nice distribution and several inferences can be made off of this distribution. The main focus of this demonstration will be the box plot. 

This data analysis shows the variability of Brady's passing yards over the years and gives the viewer information on the consistency of Brady's abilities in passing yards/game. It can be assumed that the higher the passing yards, the more consistent and improving passing over each year, i.e. the more the medians and averages are similar or improving, and the longer the career indicate a more succcesful quarterback. This is because more passing yards normally means the team has more opportunities to score in the game. Being consistent at throwing high passing yards will often mean winning more games (assuming they scored when in the positon to score) and potentiallly more championships. Finally, combining high passing yards and consistency with a long career, likely implies teams, in this case just the Patriots, want to keep their quarterback because they are successful. 

This visualization type should be used where there is information that requires analysis of distribution and range. Passing data per game can very greatly and this plot allows the viewer to have a concise view of several key values including mean, median, percentiles, maximum and minimum, etc. Furthermore, a box plot allows the user to take less space in their charts and therefore can implement more data in the chart. One last advantage of the box-plot is that it allows us to view if the data is symetrical on a year by year basis and how tight the data is. If there is large variation in the data (in this case passing yards/game) it is caught with this plotting. The subplot of the average yards a game is simply an additive measure for our visualization that shows the end user additional information. Our main focus will be the box plot.

Another graph to consider for this type of data would be a histogram or a density plot, however, those would take up more space and it would have to include a lot of different graphs to break down each individual season. One drawback of the boxplot is that it does not show an extremely detailed distribution of results. A box-plot focuses on a simple view and it can be used to quickly and efficiently compare data. However, a detailed analysis of the data would require additional graphs. This drawback does not impact the data being conveyed in our charts as there is not a ton of data per season and we are just looking for a broad understanding of Tom Brady's passing yard statistics over his career. 



Visualization Library:

For this project I will be using the Plotly graphig libraries. Plotly was created by Alex Johnson, Jack Parmer, Chris Parmer, and Matthew Sundquist in Canada in 2012. The library gives individuals graphing and analytics tools for several programming languages including Python, R, and Matlab. The library is built on top of the Plotly JavaScript library and can be used to create visualization in a Jupyter Notebook. Plotly is great for this visualization in particular because it gives the user the ability to zoom into each season, select which season is shown in the graph, allows you to hover over the season and gives statistics, etc. all without any additional coding. 

The Graphing libraries are open source and can be downloaded in either of the following manners through your terminal: 

[$ pip install plotly==4.2.1]

[$ conda install -c plotly plotly=4.2.1]

I used the chart_studio download which includes plotly 

[$ pip install chart_studio]

The general approach of the library is to create a figure (several different options including BoxPlot) and add the corresponding data to that figure. The library is declarative and is fully integrated with Jupyter. I wanted to use this library first and foremost because I had not used it before and I was interested in trying an additional visualization library outside of Matplotlib and Altair. Furthermore, as I began to learn the library, I appreciated the ease of use and the readabilty of the tutorials on the plotly site. The final product box-plot really showed the benefits of this library as well. I feel the plot is easy to read and the the ability to hover over the specific seasons showing data and the ability to select which seasons are charted are very impactful to the end user. This all was done without  additional coding, except customizatoins and therefore I thought it was an efficient and effective way to showcase the dataset.           
            

Demonstration of Box Plot in Plotly

To begin our demonstration lets import all of the software libraries we will need.

In [0]:
import pandas as pd

import numpy as np

from datetime import datetime

import chart_studio.plotly as py
import plotly.graph_objects as go
from plotly.subplots import make_subplots


Next, we will get our dataset. I chose my dataset from Pro-Football-Reference, a free statistics site for NFL players. Our dataset in particular is from the following link https://www.pro-football-reference.com/players/B/BradTo00/gamelog/ which shows the game statistics of Tom Brady's entire career. 

We will load the dataset into a dataframe and look at the beginning of the dataframe (this may take a minute or two as it is a very large dataset). 

In [2]:
url = "https://www.pro-football-reference.com/players/B/BradTo00/gamelog/"

df = pd.read_html(url, header=1)[0]

df.head()

Rk  Year        Date  G#  Week     Age  ...  Num  Pct Num.1 Pct.1 Num.2 Pct.2
0   1  2000  2000-11-23  12    13  23.112  ...  NaN  NaN   NaN   NaN   NaN   NaN
1  Rk  Year        Date  G#  Week     Age  ...  Num  Pct   Num   Pct   Num   Pct
2   2  2001  2001-09-23   2     2  24.051  ...  NaN  NaN   NaN   NaN   NaN   NaN
3   3  2001  2001-09-30   3     3  24.058  ...  NaN  NaN   NaN   NaN   NaN   NaN
4   4  2001  2001-10-07   4     4  24.065  ...  NaN  NaN   NaN   NaN   NaN   NaN

[5 rows x 51 columns]

Cleaning The Dataset:

As you can see, the dataframe created is very messy and needs a lot of cleaning up. I will update the name of the dataframe from df to cleaned_df with these changes. This is done as a precaution to indicate that the dataframe is being cleaned. I will begin by renaming some columns. I ranamed the 'Rk' column to 'Career Games Played' as it is a nice count of total games played by Brady. The other two columns I simply wanted more specificty on the column name. I then  selected the information relevant to the passing statistics replacing the previous cleaned_df. 

In [3]:
#Renamed a few Columns
df = df.rename(columns={'Rk':'Career Games Played','Date':'Game Date','Yds':'Passing Yards'})

#Passing Information

cleaned_df = df[['Career Games Played','Week','Year','Game Date','Opp','Result',
                 'Cmp','Att','Cmp%','Passing Yards','TD','Int','Rate']]

cleaned_df.head()

Career Games Played  Week  Year   Game Date  ... Passing Yards  TD  Int  Rate
0                   1    13  2000  2000-11-23  ...             6   0    0  42.4
1                  Rk  Week  Year        Date  ...           Yds  TD  Int  Rate
2                   2     2  2001  2001-09-23  ...            46   0    0  62.9
3                   3     3  2001  2001-09-30  ...           168   0    0  79.6
4                   4     4  2001  2001-10-07  ...            86   0    0  58.7

[5 rows x 13 columns]

Now that we have the passing statistics the dataframe is still a bit messy. The first problem is that there is data (here in row index 1) that is not congruent with the column titles. This is because the dataframe downloaded includes column titles for each and every year inside the dataframe. Therefore we can drop these rows using a where function combined with dropna. Secondly, in order to run statistical analysis we need to change the values of several columns to floats from strings using the asype function. 

In [4]:
#Cleaned the Df for RK

cleaned_df = cleaned_df.where(cleaned_df.values != 'Rk').dropna()


# Change Column types to Float

cleaned_df[['Cmp','Att','Cmp%','Passing Yards','TD','Int']] = cleaned_df[['Cmp','Att','Cmp%','Passing Yards','TD','Int']].astype(float)

cleaned_df.head()

Career Games Played Week  Year   Game Date  ... Passing Yards   TD  Int  Rate
0                   1   13  2000  2000-11-23  ...           6.0  0.0  0.0  42.4
2                   2    2  2001  2001-09-23  ...          46.0  0.0  0.0  62.9
3                   3    3  2001  2001-09-30  ...         168.0  0.0  0.0  79.6
4                   4    4  2001  2001-10-07  ...          86.0  0.0  0.0  58.7
5                   5    5  2001  2001-10-14  ...         364.0  2.0  0.0  93.4

[5 rows x 13 columns]

The data looks very neat and clean and we are almost ready to run some analysis. However, looking at the entire dataset for Brady, 2000 and 2008 have only one game of statistics in each season. Brady was drafted in 2000 and only played in one game in 2000. After Drew Bledsoe suffered an injury in the second week of 2001, Brady was named the starter and never lost his starting role. In 2008, after an almost perfect season in 2007 that ended in a Super Bowl loss, Brady tore his ACL and MCL in the first game of the season and was out for the rest of the season. These two datapoints will not chart well in the Box-plot methodology as both will just be a single datapoint and therefore we will remove them. We will do this with a function of having at least 10 games or more in the season. This allows the user to change the dataset and be confident in distribution of the data. IF you are using this function for another player this is still helpful as at least ten games is over half of an NFL season(16 games played) and therefore the presented chart has some consistency with players that are active on the team.

In [0]:
cleaned_df = cleaned_df.groupby('Year').filter(lambda x : len(x)>=10)

We next want to create a year list for our 'for loop' we will use in the next code set. We will also create a figure using the fig function and an open dictionary called seasons that will be used in our for loop

In [0]:
#Years

year_list = list(cleaned_df['Year'].drop_duplicates())

#Getting each seasons DataFrame
fig = go.Figure()

season={}

Creating The Figure (Box-Plot):

Our Final step is creating the boxplot for our data. We will start by creating subplots as we want to have a graph of the variabilty of cadence on the top (the boxplots) and the average cadence per game on the bottom. We set the row heights to emphasize the Box Plot as that is our main chart of analysis. We also set the shared_xaxes to be true as the data for both charts will be on the Season axis. For our boxplot we will create a 'for loop' and then explain the functionality. 

In [0]:
fig = make_subplots(rows=2, cols=1,
                    specs=[[{"secondary_y": False}],
                           [{"secondary_y": False}]], row_heights=[250,50], shared_xaxes=True, vertical_spacing=0.02)

for year in year_list:    
    season[year] = cleaned_df.loc[cleaned_df['Year'] == year]['Passing Yards']
    if int(year) % 2 == 0:
        fig.add_trace(go.Box(y=season[year], name=year, fillcolor='red', marker_color='black'))
    else:
        fig.add_trace(go.Box(y=season[year], name=year, fillcolor='blue', marker_color='black'))

The 'for loop' takes each season in the year_list and creates a new series of just the passing yards for each season.The following is an example of a dataframe for passing yards of the first five games of the 2009 season. 

In [8]:
season['2009'].head()

122    378.0
123    216.0
124    277.0
125    258.0
126    215.0
Name: Passing Yards, dtype: float64

After creating each season's passing yard dataframe, the 'for loop' then adds a boxpot to the figure with the passing data and year of the season.  A great customization feature of Plotly is the ability to customize each boxplot very easily. I decided to use an if statement to color the plots patriots colors (red and blue) based on year and marker colors to black using a very concrete set of code. Lets look at the figure:

In [9]:
fig.show()

After looking at this chart, we can see that there is no x-axis currently. We will add that later when we add the average chart. Also one can see that some boxplots have dots and fences when hovering over a sepcific season (2015 for example). Plotly defaults to this setting and it is confusing to the reader. After some digging, I found that Plotly calculates an interquartile range and the points that fall outside of that range are outliers. This is very confusing for our end viewer and we are trying to look at maximum and minimum values for each season with consistent ploting for each season.  

To alieve this problem, we create some customization with the boxpoints setting. I will change the setting in our Plotly figure to boxpoints=False. The reason for this is that I am only interested in the median, quartiles, and max and min values for each season. The chart above does not have this and some of the seasons may have interquartile range determined "fences" which makes the graph a bit more hard to read. This was a limitation with plotly's default setting but after some reading we can change the setting with the boxpoints=False function. 

In [10]:
fig.update_traces(boxpoints=False)

fig.show()

This chart has more clarity and the end user will not be confused on why some of the plots had dots and fences and some did not. We will now add the second graph below the first graph that will have the average yards per game by season. 

First we will create a new dataframe that groups our cleaned_df by Year and takes the mean. We will then reset the dataframe's index in order to use the year as our x axis value for our subplots. 

In [11]:
average_df = cleaned_df.groupby(['Year']).mean().reset_index()

average_df = average_df.round(2)

average_df.head()


Year    Cmp    Att   Cmp%  Passing Yards    TD   Int
0  2001  17.60  27.53  63.74         189.53  1.20  0.80
1  2002  23.31  37.56  61.50         235.25  1.75  0.88
2  2003  19.81  32.94  59.21         226.25  1.44  0.75
3  2004  18.00  29.62  60.19         230.75  1.75  0.88
4  2005  20.88  33.12  62.24         256.88  1.62  0.88

Creating The Figure (Average Scatter Plot):

Now that we have the dataframe we can create our second chart using the 'Year' column as our x axis and the 'Passing Yards' data (in this dataframe average/season). This chart will go in the second row of our subplots

In [12]:
fig.add_trace(go.Scatter(x=average_df['Year'], y=average_df['Passing Yards'], 
                        mode='lines', name='Average', marker_color='black'),
            row=2, col=1)

As you can see we now have x-axis data but to add clarity we now want to update the X-axes with a better angle and ensure that that each season is represented in the figure by setting tickvals = year_list. 

I then added a title to the chart 'Tom Brady Passing Yards Distribution by Season' and centered the title above the figure using x and y coordinates. I also changed the coloration of the plot background to make it easier to read and then titled the y axis of the first chart 'Yards/Game' and second chart 'Average/Game' the x axis will be 'Season.' 

In [13]:
fig.update_xaxes(tickangle=90, tickvals = year_list)

fig.update_layout(title={'text':'Tom Brady Passing Yards Distribution by Season',
                         'x':0.5, 'y':0.9,}, plot_bgcolor='rgb(243, 243, 243)')

fig.update_xaxes(title_text="Season", row=2, col=1)
fig.update_yaxes(title_text="Yards/Game", row=1, col=1)
fig.update_yaxes(title_text="Average/Game", row=2, col=1)

fig.show()

While the chart provides a nice visual, some annotation to indicate the 2008 season was not played due to injury would be helpful. After annotating the injury I also wanted to point out the Highest and Lowest Passing Yard Games in Brady's Career for our dataset. The functionality of plotly's annotation function is excellent. First I will create a dataframe of just the highest and lowest games with year and passing yards as the columns. Before we do that I will set the week and opponent data for highest and lowest passing yard game. 

In [14]:
#Find Highest and Lowest Passing Yard Data
high_row =  cleaned_df['Passing Yards'].idxmax()

high_row_df = cleaned_df.loc[[high_row]]

low_row =  cleaned_df['Passing Yards'].idxmin()

low_row_df = cleaned_df.loc[[low_row]]

#Set the week and opponent of the highest and lowest Passing Yard game
high_wk = str(high_row_df['Week'].iloc[0])
high_opp = str(high_row_df['Opp'].iloc[0])

low_wk = str(low_row_df['Week'].iloc[0])
low_opp = str(low_row_df['Opp'].iloc[0])

#Combine the two dataframes

hi_low_df = pd.concat([high_row_df,low_row_df])

hi_low_df

Career Games Played Week  Year   Game Date  ... Passing Yards   TD  Int   Rate
156                 146    1  2011  2011-09-12  ...         517.0  4.0  1.0  121.6
84                   80   17  2005  2006-01-01  ...          37.0  1.0  1.0   52.6

[2 rows x 13 columns]

Now we can see the dataframe for the highest and lowest passing yard games of Brady's career. We will now set variables that will correspond to each graph for highest and lowest game. 

In [0]:
#Set the variables x1, x2, y1, y2 to correspond with coordinates 
#on graph of year and passing yards for highest and lowest game

y1 = hi_low_df['Passing Yards'].min()

y2 = hi_low_df['Passing Yards'].max()

x1 = str(hi_low_df.loc[hi_low_df['Passing Yards'] == y1,'Year'].iloc[0])

x2 = str(hi_low_df.loc[hi_low_df['Passing Yards'] == y2,'Year'].iloc[0])

#y1 = y1.astype(int)

#y2= y2.astype(int)

After creating the annotations, we can implement the changes into our figure. I use the fig.update_layout to do this setting the x and y to the respective values in the dataframe, adding the text to be displayed and the arrow pointing to the datapoint on the chart. Adding the arrows is a limitation to Plotly as it takes a user to look at the chart and decide where best to put the annotations into the chart. Therefore the arrow data (ax and ay) are all hard coded for our chart. An advantage of this however is that the annotations are completely customizable and can be created easily by setting the x and y data for the point right into the x and y data of the chart. 

In our case we will use a customized hard coded annotation for Tom Brady for the 2008 season. Users may be tryin to use this tutorial for another player should not include the first annotation in their code. The other two annotations take the information created in our previous cells and incorporate those figures into the new figure with the x and y locations corresponding to the respective highest and lowest passing yards. As stated in the previous paragraph, if using another player the customization on the x and y location of arrow (example: ax =180, ay = 0) may need to be changed for the dataset so it is easier to read on the chart. 

Finally, Plotly boxplots allow the user to hover over each box and get data. There is an optional mode "hovermode" that allows the x, y, and closest(default) to be displayed when hovering over data. For this chart I have added in the x value when hovering over a box plot. This provides further clarity of which box plot the user is hovered over in the chart. 

In [17]:
fig.update_layout(annotations=[go.layout.Annotation(
            x='2008', y=300, xref="x", yref="y", text="Brady Injured 2008", showarrow=True, arrowhead=7, ax=0, ay=-90),
                    #The annotation above is custom to Tom Brady Only, If using another player please remove
                         
                        go.layout.Annotation(
            x=x1, y=y1, xref="x", yref="y", text="Lowest Passing Yard Game - Week {} against {}".format(low_wk,low_opp),
                            showarrow=True, arrowhead=7, ax=180, ay=0),
                    go.layout.Annotation(
            x=x2, y=y2, xref="x", yref="y", text="Highest Passing Yard Game - Week {} against {}".format(high_wk,high_opp),
                        showarrow=True, arrowhead=7, ax=180, ay=0),
                                 ], hovermode='x')


Summary of Visualization 

We can gather from the chart that Brady had several years of improved yardage to start his career and before his injury. This is shown with the median number generally increasing from 2001-2007 (albeit significant drop from 05-06). After the injury in 2008 there was a few years of lower yardage (this may not just be due to the injury) before increases and the median has stayed rather consistent in the 260-300 yards a game metric over the last ten seasons. It is interesting that the lowest and highest passing yards games were both against Miami. It is not surprising that the lowest passing yardage came in a week 17 game. This is because the Patriots have historically been in the playoffs over Brady's career and NFL teams usually rest their star players their last game in order to conserve their energy for the playoffs. In this case the Patriots were indeed in the playoff picture and Brady played sparingly at the beginning of the game. He ended the day 3 for 8 with 37 yards a touchdown and an interception. Matt Cassell, the backup quarterback played much more throwing for 168 yards and two touchdowns. 

One of the things I really like about the visual created is its clarity. Users can clearly see each season played with the distribution of data for each season. The hovermode brings further clarity to the x axis on what season's distribution is being viewed. One drawback that I had to settle with is that I found no solution to link the two subplots for hovermode as the forums I looked into stated this functionality was being worked on. This prevents the user from hovering over the boxplot graph and automatically seeing the data on the average graph (plot 2) as well. I do not think it takes away from the overall functionality of the graph, especially because we are focusing on the boxplots but it would have been a nice add-on.

A nice benefit of Plotly is that it allows the user to select specific years if they would like via the legend. This functionality can be customized to be removed but I believe it is a nice way to allow the user to select which years they would like to look at. Also Plotly allows the user to zoom into certain sections of the chart without additional code. Feel free to click on certain years on the right legend and watch as they are removed once clicked as well as zoom in on certain sections. When you are done and want to get back to the full chart just click the home symbol (reset axes) at top right of the chart and be sure to reselect all of the years. 

The data conveyed by our chart shows the following metrics by season: 

Max: Maximum Value of the Season, Q3: Third Quartile, Median (2nd Quartile): Value in wich the set is divided into halves (Middle Value or average of two middle values), Q1: First Quartile, Min: Minimum Value of Season. 

This data analysis shows the variability of Brady's passing yards over the years and gives the viewer information on the consistency of Brady's abilities in passing yards/game. Brady has had arguably the best career in the history of NFL quarterbacks. A portion of his successes is displayed in this chart. The passing yard improvement throughout his career, the more consistency over each year, i.e. the medians hovering from 260-300 yards over the last decade, and the long career of almost two decades depict a career full of tremendous accomplishments including 6 Super Bowl Championships. This visualization is just one summary of those accomplishments. 

Rules for Writing and Sharing Computational Analysis

The following are rules I adhered to while writing and sharing my computational analysis:

Rule 1: Tell a story for an audience - by using exploratory text, I was able to tell the story of Tom Brady's passing game statistics. I began the notebook explaining what we would be doing, the middle of the notebook had the steps to achieve the results and the summary of the visualization analyzed the results. 

Rule 2: Document the process, not just the results- during the meat of the notebook I described each step including cleaning the dataframe and the problems that we ran into after each step in order to show the entire process of cleaning the dataframe. I also created and updated the figure several times to show the steps at improving the visualization with comments along the way to help guide the reader.  

Rule 3: Use Cell Divisions to Make Steps Clear - the article indicated that a problem with some notebooks is that they can be filled with short fragments that are hard to follow. I have separated my cells into clear and concise steps in cleaning the dataset, creating the charts, adding annotations, etc. I also did not have any lines of cells that are longer than 40 lines which makes the code easier to read. I also added some comments in the code itself which I feel helps the user follow the lines of code more clearly. 

Rule 8: Share and explain your data- I used data from a free, publicly available website. This allows me to share the data alongside my notebook and helps the user to interpret the results more easily. I also described the dataset given in the beginning of the notebook. 

Chadwick Schlett : Graduate Student : University of Michigan Master of Applied Datascience 